File: sdk/js/src/guard.ts
/**
 * Guard / verifier helpers.
 *
 * A guard is a wrapper-side function the core invokes via `guard.execute` to
 * decide whether a prompt / tool call / output should be allowed, denied, or
 * should trip the safety wire (immediate stop).
 *
 * A verifier is a wrapper-side function the core invokes via `verifier.execute`
 * after a tool produced a result, returning `{passed, summary}`.
 *
 * Stage-specific signatures (sync or async — both are supported):
 *
 * ```ts
 * inputGuard('no_secrets', (input) => ({ decision: 'deny', reason: '...' }));
 * toolCallGuard('fs_root_only', (toolName, args) => ({ decision: 'allow' }));
 * outputGuard('pii_redactor', (output) => ({ decision: 'allow' }));
 * verifier('non_empty', (toolName, args, result) => ({ passed: true }));
 * ```
 */

export const STAGE_INPUT = 'input' as const;
export const STAGE_TOOL_CALL = 'tool_call' as const;
export const STAGE_OUTPUT = 'output' as const;

export type GuardStage = typeof STAGE_INPUT | typeof STAGE_TOOL_CALL | typeof STAGE_OUTPUT;

export type GuardDecision = 'allow' | 'deny' | 'tripwire';

const VALID_DECISIONS: ReadonlySet<GuardDecision> = new Set(['allow', 'deny', 'tripwire']);

export interface GuardResult {
  decision: GuardDecision;
  reason?: string;
}

export type InputGuardFn = (input: string) => GuardResult | Promise<GuardResult>;
export type ToolCallGuardFn = (
  toolName: string,
  args: Record<string, unknown>,
) => GuardResult | Promise<GuardResult>;
export type OutputGuardFn = (output: string) => GuardResult | Promise<GuardResult>;

export interface GuardDefinition {
  readonly name: string;
  readonly stage: GuardStage;
  /** Wrapper-side handler dispatched from `guard.execute`. */
  readonly call: (params: {
    input?: string;
    toolName?: string;
    args?: Record<string, unknown>;
    output?: string;
  }) => Promise<GuardResult>;
}

function normalise(result: GuardResult): GuardResult {
  if (!VALID_DECISIONS.has(result.decision)) {
    return {
      decision: 'deny',
      reason: `invalid guard decision ${JSON.stringify(result.decision)}: ${result.reason ?? ''}`,
    };
  }
  return { decision: result.decision, reason: result.reason ?? '' };
}

/** Build an input-stage guard. */
export function inputGuard(name: string, fn: InputGuardFn): GuardDefinition {
  return {
    name,
    stage: STAGE_INPUT,
    call: async ({ input = '' }) => normalise(await fn(input)),
  };
}

/** Build a tool-call-stage guard. */
export function toolCallGuard(name: string, fn: ToolCallGuardFn): GuardDefinition {
  return {
    name,
    stage: STAGE_TOOL_CALL,
    call: async ({ toolName = '', args = {} }) => normalise(await fn(toolName, args)),
  };
}

/** Build an output-stage guard. */
export function outputGuard(name: string, fn: OutputGuardFn): GuardDefinition {
  return {
    name,
    stage: STAGE_OUTPUT,
    call: async ({ output = '' }) => normalise(await fn(output)),
  };
}

/** Internal: project a guard into the wire format used by `guard.register`. */
export function guardToProtocolDict(def: GuardDefinition): Record<string, unknown> {
  return { name: def.name, stage: def.stage };
}

// ---------------------------------------------------------------------------
// Verifiers (also wrapper-side; live next to guards because they share shape)
// ---------------------------------------------------------------------------

export interface VerifierResult {
  passed: boolean;
  summary?: string;
}

export type VerifierFn = (
  toolName: string,
  args: Record<string, unknown>,
  result: string,
) => VerifierResult | Promise<VerifierResult>;

export interface VerifierDefinition {
  readonly name: string;
  readonly call: (params: {
    toolName: string;
    args: Record<string, unknown>;
    result: string;
  }) => Promise<VerifierResult>;
}

/** Build a verifier definition. */
export function verifier(name: string, fn: VerifierFn): VerifierDefinition {
  return {
    name,
    call: async ({ toolName, args, result }) => {
      const out = await fn(toolName, args, result);
      return { passed: !!out.passed, summary: out.summary ?? '' };
    },
  };
}

/** Internal: project a verifier into the wire format. */
export function verifierToProtocolDict(def: VerifierDefinition): Record<string, unknown> {
  return { name: def.name };
}

File: sdk/python/ai_agent/guard.py
"""Guard decorators: ``@input_guard``, ``@tool_call_guard``, ``@output_guard``.

A guard is a wrapper-side function the core invokes via ``guard.execute`` to
decide whether a prompt / tool call / output should be allowed, denied, or
should trip the safety wire (immediate stop).

Each guard returns a 2-tuple ``(decision, reason)`` where ``decision`` is one
of ``"allow"``, ``"deny"``, ``"tripwire"``. ``reason`` is a short human-
readable explanation that the core may surface to the user.

Stage-specific signatures (sync or async — both are supported):

    @input_guard(name="no_secrets")
    def check_input(input: str) -> tuple[str, str]: ...

    @tool_call_guard(name="fs_root_only")
    def check_tool_call(tool_name: str, args: dict) -> tuple[str, str]: ...

    @output_guard(name="pii_redactor")
    def check_output(output: str) -> tuple[str, str]: ...
"""

from __future__ import annotations

import asyncio
from dataclasses import dataclass, field
from typing import Any, Awaitable, Callable

# Stage identifiers (must match pkg/protocol/methods.go).
STAGE_INPUT = "input"
STAGE_TOOL_CALL = "tool_call"
STAGE_OUTPUT = "output"

VALID_DECISIONS = frozenset({"allow", "deny", "tripwire"})


GuardCallable = Callable[..., "tuple[str, str] | Awaitable[tuple[str, str]]"]


@dataclass
class GuardDefinition:
    """SDK-side record of a registered guard."""

    name: str
    stage: str
    func: GuardCallable = field(repr=False)
    is_coroutine: bool = field(repr=False, default=False)

    async def call(self, *, input: str = "", tool_name: str = "",
                   args: Any | None = None, output: str = "") -> tuple[str, str]:
        """Dispatch the wrapped function with the right kwargs for the stage.

        Sync functions run in the default executor to avoid blocking the
        JSON-RPC reader loop.
        """

        if self.stage == STAGE_INPUT:
            kwargs = {"input": input}
        elif self.stage == STAGE_TOOL_CALL:
            kwargs = {"tool_name": tool_name, "args": args or {}}
        elif self.stage == STAGE_OUTPUT:
            kwargs = {"output": output}
        else:  # pragma: no cover — guarded at decoration time.
            raise ValueError(f"unknown guard stage: {self.stage}")

        if self.is_coroutine:
            decision, reason = await self.func(**kwargs)
        else:
            loop = asyncio.get_running_loop()
            decision, reason = await loop.run_in_executor(
                None, lambda: self.func(**kwargs)
            )

        if decision not in VALID_DECISIONS:
            # Fail-closed: unknown decisions are treated as deny.
            return "deny", f"invalid guard decision {decision!r}: {reason}"
        return decision, reason

    def to_protocol_dict(self) -> dict[str, Any]:
        return {"name": self.name, "stage": self.stage}


def _make_decorator(stage: str):
    def decorator_factory(*, name: str | None = None):
        def decorator(func: Callable[..., Any]) -> Callable[..., Any]:
            defn = GuardDefinition(
                name=name or func.__name__,
                stage=stage,
                func=func,
                is_coroutine=asyncio.iscoroutinefunction(func),
            )
            setattr(func, "__ai_agent_guard__", defn)
            return func

        return decorator

    return decorator_factory


input_guard = _make_decorator(STAGE_INPUT)
tool_call_guard = _make_decorator(STAGE_TOOL_CALL)
output_guard = _make_decorator(STAGE_OUTPUT)


def get_guard_definition(obj: Any) -> GuardDefinition | None:
    """Return the :class:`GuardDefinition` attached to ``obj`` by a guard
    decorator, or ``None`` if absent."""

    return getattr(obj, "__ai_agent_guard__", None)


__all__ = [
    "GuardCallable",
    "GuardDefinition",
    "STAGE_INPUT",
    "STAGE_TOOL_CALL",
    "STAGE_OUTPUT",
    "get_guard_definition",
    "input_guard",
    "output_guard",
    "tool_call_guard",
]
